container/lxd/lxd.go

```diff
@@ -36,26 +46,40 @@ const lxdDefaultProfileName = "default"
 type containerManager struct {
 	modelUUID string
 	namespace instance.Namespace
-	// A cached client.
-	client *lxdclient.Client
+	server    lxd.ContainerServer
 	// a host machine's availability zone
 	availabilityZone string
 }
 
 // containerManager implements container.Manager.
 var _ container.Manager = (*containerManager)(nil)
 
-func ConnectLocal() (*lxdclient.Client, error) {
-	cfg := lxdclient.Config{
-		Remote: lxdclient.Local,
+type Device map[string]string
+type Devices map[string]Device
+
+func lxdSocketPath() string {
+	// LXD socket is different depending on installation method
+	// We prefer upstream's preference of snap installed LXD
+	debianSocket := filepath.FromSlash("/var/lib/lxd")
+	snapSocket := filepath.FromSlash("/var/snap/lxd/common/lxd")
+	path := os.Getenv("LXD_DIR")
+	if path != "" {
+		logger.Debugf("Using environment LXD_DIR as socket path: %q", path)
+	} else {
+		if _, err := os.Stat(snapSocket); err == nil {
+			logger.Debugf("Using LXD snap socket: %q", snapSocket)
+			path = snapSocket
+		} else {
+			logger.Debugf("LXD snap socket not found, falling back to debian socket: %q", debianSocket)
+			path = debianSocket
+		}
 	}
+	return filepath.Join(path, "unix.socket")
+}
 
-	cfg, err := cfg.WithDefaults()
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
+func ConnectLocal() (lxd.ContainerServer, error) {
+	client, err := lxd.ConnectLXDUnix(lxdSocketPath(), &lxd.ConnectionArgs{})
```

Comment:
yay for not setting an env var that has a side-effect of working for the lxd client